irc/metadata.go

```diff
@@ -0,0 +1,257 @@
+	}
+
+	copy := make(MetadataStore)
+	maps.Copy(copy, data)
+
+	return true, copy
+}
+
+func metadataDelete(server *Server, session *Session, target string, key string) {
+	targetChannel := server.channels.Get(target)
+	targetClient := server.clients.Get(target)
+
+	if targetChannel != nil {
+		targetChannel.stateMutex.Lock()
+		defer targetChannel.stateMutex.Unlock()
+
+		delete(targetChannel.metadata, key)
+		return
+	}
+
+	if targetClient != nil {
+		targetClient.stateMutex.Lock()
+		defer targetClient.stateMutex.Unlock()
+
+		delete(targetClient.metadata, key)
+		return
+	}
+
+	notifySubscribers(server, session, target, key, "")
+}
+
+func metadataClearTarget(server *Server, target string) (bool, MetadataStore) {
+	targetChannel := server.channels.Get(target)
+	targetClient := server.clients.Get(target)
+
+	if targetChannel != nil {
+		targetChannel.stateMutex.Lock()
+		defer targetChannel.stateMutex.Unlock()
+
+		oldMap := targetChannel.metadata
+		targetChannel.metadata = make(MetadataStore)
+		return true, oldMap
+	}
+
+	if targetClient != nil {
+		targetClient.stateMutex.Lock()
+		defer targetClient.stateMutex.Unlock()
+
+		oldMap := targetClient.metadata
+		targetClient.metadata = make(MetadataStore)
+		return true, oldMap
+	}
+
+	return false, nil
+}
+
+var metadataEvilCharsRegexp = regexp.MustCompile("[^A-Za-z0-9_./:-]+")
+
+func metadataKeyIsEvil(key string) bool {
+	key = strings.TrimSpace(key) // just in case
+
+	return len(key) == 0 || // key needs to contain stuff
+		key[0] == ':' || // key can't start with a colon
+		metadataEvilCharsRegexp.MatchString(key) // key can't contain the stuff it can't contain
+}
+
+func metadataCanIEditThisKey(client *Client, target string, _ string) bool {
+	if !metadataCanIEditThisTarget(client, target) { // you can't edit keys on targets you can't edit.
+		return false
+	}
+
+	// todo: we don't actually do anything regarding visibility yet so there's not much to do here
+
+	return true
+}
+
+func metadataCanIEditThisTarget(client *Client, target string) bool {
+	if !metadataCanISeeThisTarget(client, target) { // you can't edit what you can't see. a wise man told me this once
+		return false
+	}
+
+	if client.HasRoleCapabs("sajoin") { // sajoin opers can do whatever they want
+		return true
+	}
+
+	if target == client.nick { // your right to swing your fist ends where my nose begins
+		return true
+	}
+
+	// if you're a channel operator, knock yourself out
+	channel := client.server.channels.Get(target)
+	if channel != nil && channel.ClientIsAtLeast(client, modes.Operator) {
+		return true
+	}
+
+	return false
+}
+
+func metadataCanISeeThisTarget(client *Client, target string) bool {
+	if client.HasRoleCapabs("sajoin") { // sajoin opers can do whatever they want
+		return true
+	}
+
+	// check if this is a secret channel and, if so, check if they can see it
+	channel := client.server.channels.Get(target)
+	if channel != nil && channel.flags.HasMode(modes.Secret) && !channel.hasClient(client) {
```

Comment:
I think the client needs to be a member to view channel metadata whether or not the channel is secret (e.g. consider a `+i` channel that the client is not allowed to be in)